src/clients/FtpClient.h

```diff
@@ -99,11 +99,17 @@ class DataChannel: public Ftp::Channel
 
     void addr(const Ip::Address &addr); ///< import host and port
 
+    /// updates counters after this number of bytes have been added to readBuf
+    void appended(size_t size);
+
 public:
     MemBuf *readBuf;
     char *host;
     unsigned short port;
     bool read_pending;
+
+    /// amount of message payload/body received so far
+    int64_t payloadSeen = 0;
```

Comment:
If possible:

```suggestion
    uint64_t payloadSeen = 0;
```

Reply:
payloadSeen is compared with Gateway::theSize, which is signed. I would convert both or none.

Reply:
payloadSeen is maintained using unsigned (size_t) `size`. Thus, we cannot make an "only used with signed types" argument. Moreover, the only comparison with signed theSize is explicitly preconditioned on theSize being positive. Thus, I think payloadSeen should be unsigned from birth.

Said that, theSize should be converted to std::optional<uint64_t>. Unfortunately, theSize seems to participate in many _inconsistent_ conditions:

```c++
        if (ftpState->theSize == 0)
    if (theSize <= 0)
            ftpState->mdtm <= 0 && ftpState->theSize < 0)

    if (request->method == Http::METHOD_HEAD && (flags.isdir || theSize != -1)) {
    if (ftpState->request->method == Http::METHOD_HEAD && (ftpState->flags.isdir || ftpState->theSize != -1)) {
```

We should change theSize type in a dedicated PR. For now, let's just cast theSize to an unsigned integer where we compare it with payloadSeen.

Reply:
Done.